cpp/src/arrow/compute/kernels/scalar_string_test.cc

```diff
@@ -724,7 +724,7 @@ TYPED_TEST(TestStringKernels, IsUpperAscii) {
                    "[false, null, false, true, true, false, false]");
 }
 
-TYPED_TEST(TestStringKernels, MatchSubstring) {
+TYPED_TEST(TestBinaryKernels, MatchSubstring) {
```

Comment:
For any test that now includes binary values, we need to test invalid UTF-8.

Reply:
Ok, so I noticed that there are not binary (invalid UTF8) tests for string functions. As a starting point, I chose `binary_length` as an "easy" function to test,
```c++
// These cases pass successfully
this->CheckUnary("binary_length", "[\"\x84\xf0\x2f\", \"\xff\x9b\xc3\xbb\"]", this->offset_type(), "[3, 4]");
```
but the JSON parser errors out in many cases (e.g., `\x84\x00`, `\x84\x01`) with one of the following messages:
```c++
Invalid: Expected string or null, got JSON type 6
Invalid: JSON parse error at offset 3: Invalid escape character in string.
Invalid: JSON parse error at offset 3: Missing a closing quotation mark in string.
...
```
I suspect that the successful cases are valid UTF-8 (need to check), but the failed tests are definitely not valid UTF-8.

Reply:
Will test using `MakeScalar` and bypassing the JSON parser.

Reply:
I suspect the issue is that JSON requires inputs to be in UTF-8 or another encoding of Unicode. See for instance RapidJSON's documentation (which is what we use): https://rapidjson.org/md_doc_encoding.html

So for such tests, you may have to manually construct input data. (Or it seems like the validation can be turned off in RapidJSON.) A helper that takes a vector<std::string> and gives back an array might suffice.

Reply:
Up to now, the only limitation with non-encoded binary input for all the [Large]Binary/StringTypes is that a null byte (`0x00`) cannot appear in the input because it is treated as the null-terminator for `string_view`. More details on this are in the [JIRA comments](https://issues.apache.org/jira/browse/ARROW-13879).